fuzzer/src/main.rs

```diff
@@ -0,0 +1,108 @@
+use parity_scale_codec::{Encode, Decode, Compact};
+use std::collections::{BTreeMap, BTreeSet, VecDeque, LinkedList, BinaryHeap};
+use honggfuzz::fuzz;
+use bitvec::{vec::BitVec, cursor::BigEndian};
+
+#[derive(Encode, Decode)]
+pub struct MockStruct{
+	vec_u: Vec<u8>
+}
+
+#[derive(Encode, Decode)]
+pub enum MockEnum {
+	Empty,
+	Unit(u32),
+	UnitVec(Vec<u8>),
+	Complex {
+		data: Vec<u32>,
+		map: BTreeMap<[u8; 32], Vec<u8>>,
+		string: String,
+	},
+	Mock(MockStruct),
+	NestedVec(Vec<Vec<Vec<Vec<Vec<Vec<Vec<Vec<Option<u8>>>>>>>>>),
+}
+
+macro_rules! fuzz_types {
+	(
+		$data:ident;
+		$first:ty,
+		$( $rest:ty, )*
+	) => {
+		fuzz_types! {
+			@INTERNAL
+			$data;
+			1u8;
+			{ $first; 0u8 }
+			$( $rest, )*
+		}
+	};
+	(@INTERNAL
+		$data:ident;
+		$counter:expr;
+		{ $( $parsed:ty; $index:expr ),* }
+		$current:ty,
+		$( $rest:ty, )*
+	) => {
+		fuzz_types! {
+			@INTERNAL
+			$data;
+			$counter + 1u8;
+			{ $current; $counter $(, $parsed; $index )* }
+			$( $rest, )*
+		}
+	};
+	(@INTERNAL
+		$data:ident;
+		$counter:expr;
+		{ $( $parsed:ty; $index:expr ),* }
+	) => {
+		let num = $counter;
+		$(
+			if $data[0] % num == $index {
+				let _ = <$parsed>::decode(&mut &$data[1..]);
```

Comment:
would it make sense to add a test for min_encoded_len here ? like if result is success then we check min_encoded_len is coherent

Reply:
You can add this.